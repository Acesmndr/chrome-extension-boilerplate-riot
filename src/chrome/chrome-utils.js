/* global chrome */
const clearCache = () => {
  return new Promise((resolve) => {
    chrome.storage.local.clear(() => {
      resolve();
    });
  });
}

/* fetch returns a promise which resolves with the data from storage */
const fetch = (getWhat) => {
  return new Promise((resolve) => {
    chrome.storage.local.get(getWhat,(data) => {
      resolve(data);
    });
  });
}

/* triggers a chrome notification */
const notify = (params) => {
  return new Promise((resolve) => {
    chrome.notifications.create({
      type: 'basic',
      title: params.title || 'Chrome Extension',
      message: params.message || '',
      buttons: params.buttons ? [{ title: 'Yes' }, { title: 'No' }] : [],
      requireInteraction: false,
      iconUrl: 'assets/img/icon.png',
    }, resolve);
  });
}

/* dispatch a message to listeners in background or popup */
const sendMessage = (msg) => {
  return new Promise((resolve) => {
    chrome.runtime.sendMessage(msg);
    resolve();
  });
}

/* sets the browser action icon */
const setBadgeIcon = (iconPath) => {
  chrome.browserAction.setIcon({path: iconPath});
}

/* 
  sets the browser action badge text
  only four characters are visible
*/
const setBadgeText = (text = 'ext') => {
  chrome.browserAction.setBadgeText({
    text: text,
  });
}

/*
  sets up the context menu of the browser action
*/
const setupContextMenu = (reset) => {
  chrome.contextMenus.removeAll(() => {
    chrome.contextMenus.create({ id: 'reset_extension', title: 'Reset extension', contexts: ['browser_action'] });
    chrome.contextMenus.onClicked.addListener((info) => {
      if (info.menuItemId === 'reset_extension') {
        reset();
      }
    });
  });
}

/* on browser startup returns a promise */
const setupStartup = () => {
  return new Promise((resolve, reject) => {
    chrome.runtime.onStartup.addListener(() => {
      resolve();
    });
  });
}

/* stores a data in the local storage */
const store = (dataObj) => {
  return new Promise((resolve, reject) => {
    chrome.storage.local.set(dataObj, resolve);
  });
}

export {
  clearCache,
  fetch,
  notify,
  sendMessage,
  setBadgeIcon,
  setBadgeText,
  setupContextMenu,
  setupStartup,
  store,
}

