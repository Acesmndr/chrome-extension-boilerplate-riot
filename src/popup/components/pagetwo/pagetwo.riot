<pagetwo>
	<div class='card'>
		<div class='content'>
			<div class='form'>
				<div class='title'>To Do</div>
				<ul class='todolist'>
					<li each = { (todoItem, i) in (state.todolist || []) } onclick={ removeItem } itemNo={i}>{ todoItem }</li>
				</ul>
				<img src='assets/img/empty.svg' class='empty' if={ !state.todolist || !state.todolist.length } />
				<input type='text' required placeholder='Add item' autofocus/>
				<input type='button' class='add-btn' value='+' onclick={ addItem } />
			</div>
		</div>
		<div class="content right-align buttons">
				<input type='button' class='clear' onclick={ clearAll } value='Clear All'/>
			<input type='button' onclick={ goToFirstPage } value='Return to page 1'/>
		</div>
	</div>
	<style type='scss'>
		@import './src/popup/components/pagetwo/pagetwo';
	</style>
	<script>
		import script from './pagetwo.js';
		export default script;
	</script>
</pagetwo>
