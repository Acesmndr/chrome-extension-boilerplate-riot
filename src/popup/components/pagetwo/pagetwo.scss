@import '../../style/colorpalette';

.form {
  font-size: 12px;
  .title {
    margin-bottom: 24px;
    color: $simply-grey;
    text-align: center;
    font-size: 22px;
  }
  .empty {
    margin-bottom: 32px;
    width: 320px;
  }
  input[type='text'] {
    padding: 10px;
    width: 260px;
    border: 1px solid $simply-grey;
    border-radius: 2px 0 0 2px;
    font-size: 16px;
  }
  .add-btn {
    background: $simply-green;
    border-radius: 0 2px 2px 0;
    margin: 0;
    font-size: 24px;
    line-height: 24px;
    padding: 5px 12px 11px 10px;
  }
}

.todolist {
  li {
    font-size: 20px;
    margin-bottom: 24px;
    color: $simply-grey;
    &:hover {
      text-decoration: line-through;
    }
  }  
}

.right-align.buttons {
  margin-top: 32px;
}
