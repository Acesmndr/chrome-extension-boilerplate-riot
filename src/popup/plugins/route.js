import { mount, unmount } from 'riot';

export default function(component) {
  component.route = (page, options = {}) => {
    unmount(document.getElementById('root'), true);
    mount(document.getElementById('root'), options, page);
  };

  return component;
}