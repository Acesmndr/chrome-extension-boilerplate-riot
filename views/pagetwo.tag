<pagetwo>
	<div class='card'>
		<div class='content'>
			<form>
				<label for='name-input'>Enter a todo</label>
				<input type='text' id='name-input'/>
				<input type='button' value='Add' />
			</form>
		</div>
		<div class="content right-align">
			<input type='button' onclick={ pageRedirect } value='Return to page 1'/>
		</div>
	</div>
	<style type='scss'>
		@import 'styles/pagetwo';
	</style>
	<script src="presenters/pagetwo.js"></script>
</pagetwo>