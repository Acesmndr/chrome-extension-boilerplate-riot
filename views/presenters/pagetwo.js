this.goToFirstPage = () => {
  riot.mount('.main-body', 'pageone');
}

this.on('mount', () => {
  this.mixin('messageMixin');
  this.setupListener(this);
  chrome.runtime.sendMessage({
    type: 'getDataFromBackground',
  });
});

this.on('unmount', () => {
  console.log('Blank got unmounted');
});