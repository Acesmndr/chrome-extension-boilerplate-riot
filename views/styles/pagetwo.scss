@import './colorpalette';
input#name-input {
  padding: 10px;
  font-size: 16px;
}
form {
  padding: 16px;
  font-size: 16px;
}