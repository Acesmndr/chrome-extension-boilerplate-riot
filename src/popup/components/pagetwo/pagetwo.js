import helper from '../../helper';
import { sendMessage } from '../../../chrome/chrome-utils';

export default {
  state: {
    todolist: [],
    containsItem: false,
  },
  addItem() {
    const item = this.$('input[type="text"]').value;
    if (item === '') {
      return;
    }
    sendMessage({
      type: 'saveDataInBackground',
      data: { todolist: [...this.state.todolist, item] },
    });
    this.$('input[type="text"]').value = '';
  },
  removeItem(e) {
    let itemList = this.state.todolist;
    debugger;
    itemList.splice(Number(e.target.getAttribute('itemNo')), 1);
    sendMessage({
      type: 'saveDataInBackground',
      data: { todolist: itemList },
    });
  },
  clearAll() {
    sendMessage({
      type: 'removeAllDataInBackground',
    })
  },
  goToFirstPage() {
    helper.route('pageone');
  },
  onBeforeMount(props, state) {
    sendMessage({
      type: 'getDataFromBackground',
      query: 'todolist'
    });
  },
  onBeforeUpdate(props, state) {
    console.log(props, state);
    state = {
      todolist: props.todolist || [],
      containsItem: props.containsItem || false,
    }
  }
}