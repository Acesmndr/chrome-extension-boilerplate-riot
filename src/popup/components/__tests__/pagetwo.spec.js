import * as riot from 'riot';
import pagetwo from '../pagetwo/pagetwo.riot';
import route from '../../plugins/route';
import * as chromeUtils from '../../../chrome/chrome-utils';

riot.register('pagetwo', pagetwo);
riot.install(route);

describe('Page two specs', () => {
  const pageElement = document.createElement('div');
  const mockFn = jest.fn();
  let mountedComponent;

  beforeAll(() => {
    pageElement.id = 'root';
    document.body.appendChild(pageElement);

    chromeUtils.sendMessage = mockFn;
    mountedComponent = riot.mount(pageElement, {}, 'pagetwo');
  });

  afterEach(() => {
    chromeUtils.sendMessage.mockReset();
  });

  afterAll(() => {
    chromeUtils.sendMessage.mockRestore();
  })


  it('should fetch the data in onBeforeMount function', () => {
    expect(chromeUtils.sendMessage.mock.calls.length).toBe(1);
    expect(chromeUtils.sendMessage.mock.calls[0]).toEqual([{
      type: 'getDataFromBackground',
      query: 'todolist'
    }]);
  });
  
  it('should mount the tag', () => {
    expect(document.querySelector('#root')).toMatchSnapshot();
  });


  it('should have a title text', () => {
    expect(document.querySelector('.form .title').textContent).toBe('To Do');
  });

  it('should have a input text field', () => {
    expect(document.querySelectorAll('input[type="text"]').length).toBe(1);
  });

  it('should add an item on todo list', () => {
    document.querySelector('input[type="text"]').value = "wonderwoman";
    const addButton = document.querySelector('input[type="button"][value="+"]');
    addButton.click();
    
    expect(chromeUtils.sendMessage.mock.calls.length).toBe(1);
    expect(chromeUtils.sendMessage.mock.calls[0]).toEqual([{
      type: 'saveDataInBackground',
      data: { todolist: ['wonderwoman'] },
    }]);
  });

  it('should remove an item on todo list', () => {
    mountedComponent[0].state = {
      todolist: ['batman', 'superman'],
      containsItem: true,
    }
    mountedComponent[0].removeItem({
      target: {
        getAttribute: (attr) => {
          if(attr === 'itemNo') {
            return '1';
          }
          return undefined;
        }
      }
    });

    expect(chromeUtils.sendMessage.mock.calls.length).toBe(1);
    expect(chromeUtils.sendMessage.mock.calls[0]).toEqual([{
      type: 'saveDataInBackground',
      data: { todolist: ['batman'] },
    }]);
  });

  it('should send a clear request on clearAll button click', () => {
    const clearButton = document.querySelector('input[type="button"][value="Clear All"]');
    clearButton.click();
    
    expect(chromeUtils.sendMessage.mock.calls.length).toBe(1);
    expect(chromeUtils.sendMessage.mock.calls[0]).toEqual([{
      type: 'removeAllDataInBackground',
    }]);
  });

  it('should route to second page when route button is pressed', () => {
    const routeButton = document.querySelector('input[type="button"][value="Return to page 1"]');
    const mockFn = jest.fn();

    riot.mount = mockFn;
    routeButton.click();

    expect(riot.mount.mock.calls.length).toBe(1);
    expect(riot.mount.mock.calls[0]).toEqual([pageElement, {}, "pageone"]);

    riot.mount.mockRestore();
  })
});