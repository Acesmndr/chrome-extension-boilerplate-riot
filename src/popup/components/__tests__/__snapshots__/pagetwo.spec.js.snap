// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page two specs should mount the tag 1`] = `
<div
  class="main-body"
  data-is="pagetwo"
>
  <div
    class="card"
  >
     
    <div
      class="content"
    >
       
      <div
        class="form"
      >
         
        <div
          class="title"
        >
          To Do
        </div>
         
        <ul
          class="todolist"
        >
           
          <li>
            batman
          </li>
          <li>
            superman
          </li>
          
           
        </ul>
         
        
         
        <input
          autofocus=""
          placeholder="Add item"
          ref="nameInput"
          required=""
          type="text"
        />
         
        <input
          class="add-btn"
          type="button"
          value="+"
        />
         
      </div>
       
    </div>
     
    <div
      class="content right-align buttons"
    >
       
      <input
        type="button"
        value="Clear All"
      />
       
      <input
        type="button"
        value="Return to page 1"
      />
       
    </div>
     
  </div>
</div>
`;
